pkg/meta/repodb/boltdb-wrapper/boltdb_wrapper.go

```diff
@@ -1976,28 +1878,309 @@ func getUserBookmarks(ctx context.Context, transaction *bbolt.Tx) []string {
+		userid   = localCtx.GetUsernameFromContext(acCtx)
+		userdb   = transaction.Bucket([]byte(bolt.UserDataBucket))
 	)
 
-	if userid == "" {
+	if userid == "" || userdb == nil {
 		return []string{}
 	}
 
-	if userBucket == nil {
+	mdata := userdb.Get([]byte(userid))
+	if mdata == nil {
 		return []string{}
 	}
 
-	mdata := userBucket.Get([]byte(bolt.BookmarkedReposKey))
-	if mdata == nil {
+	if err := json.Unmarshal(mdata, &userData); err != nil {
 		return []string{}
 	}
 
-	if err := json.Unmarshal(mdata, &bookmarkedRepos); err != nil {
-		return []string{}
+	return userData.BookmarkedRepos
+}
+
+func (bdw *DBWrapper) SetUserGroups(ctx context.Context, groups []string) error {
+	acCtx, err := localCtx.GetAccessControlContext(ctx)
+	if err != nil {
+		return err
+	}
+
+	userid := localCtx.GetUsernameFromContext(acCtx)
+
+	if userid == "" {
+		// empty user is anonymous
+		return zerr.ErrUserDataNotAllowed
+	}
+
+	err = bdw.DB.Update(func(tx *bbolt.Tx) error { //nolint:varnamelen
+		var userData repodb.UserData
+
+		err := bdw.getUserData(userid, tx, &userData)
+		if err != nil && !errors.Is(err, zerr.ErrUserDataNotFound) {
+			return err
+		}
+
+		userData.Groups = append(userData.Groups, groups...)
+
+		err = bdw.setUserData(userid, tx, userData)
+
+		return err
+	})
+
+	return err
+}
+
+func (bdw *DBWrapper) GetUserGroups(ctx context.Context) ([]string, error) {
+	userData, err := bdw.GetUserData(ctx)
+
+	return userData.Groups, err
+}
+
+func (bdw *DBWrapper) UpdateUserAPIKeyLastUsed(ctx context.Context, hashedKey string) error {
+	acCtx, err := localCtx.GetAccessControlContext(ctx)
+	if err != nil {
+		return err
+	}
+
+	userid := localCtx.GetUsernameFromContext(acCtx)
+
+	if userid == "" {
+		// empty user is anonymous
+		return zerr.ErrUserDataNotAllowed
+	}
+
+	err = bdw.DB.Update(func(tx *bbolt.Tx) error { //nolint:varnamelen
+		var userData repodb.UserData
+
+		err := bdw.getUserData(userid, tx, &userData)
+		if err != nil {
+			return err
+		}
+
+		apiKeyDetails := userData.APIKeys[hashedKey]
+		apiKeyDetails.LastUsed = time.Now()
+
+		userData.APIKeys[hashedKey] = apiKeyDetails
+
+		err = bdw.setUserData(userid, tx, userData)
+
+		return err
+	})
+
+	return err
+}
+
+func (bdw *DBWrapper) AddUserAPIKey(ctx context.Context, hashedKey string, apiKeyDetails *repodb.APIKeyDetails) error {
+	acCtx, err := localCtx.GetAccessControlContext(ctx)
+	if err != nil {
+		return err
+	}
+
+	userid := localCtx.GetUsernameFromContext(acCtx)
+	if userid == "" {
+		// empty user is anonymous
+		return zerr.ErrUserDataNotAllowed
+	}
+
+	err = bdw.DB.Update(func(transaction *bbolt.Tx) error {
+		var userData repodb.UserData
+
+		apiKeysbuck := transaction.Bucket([]byte(bolt.UserAPIKeysBucket))
+		if apiKeysbuck == nil {
+			return zerr.ErrBucketDoesNotExist
+		}
+
+		err := apiKeysbuck.Put([]byte(hashedKey), []byte(userid))
+		if err != nil {
+			return fmt.Errorf("repoDB: error while setting userData for identity %s %w", userid, err)
```

Comment:
Aren't these errors supposed to be defined in the errors package?

Reply:
not really, because I don't need to treat them outside, I fixed by just returning boltdb errors

Reply:
What do you mean by not needing to threat them outside?

Reply:
to handle them, sorry, like errors.Is()